cloudwatch/helpers_test.go

```diff
@@ -0,0 +1,36 @@
+package cloudwatch
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestTagKeysToMap(t *testing.T) {
+	// Testable values. Purposely "messed up" - they should all parse out OK.
+	values := " key1 =value , key2=value2, key3= value3 ,key4=, key5  = v5,,key7==value7, k8, k9,key1=value1,space key = space value"
+	// The values above should return a map like this.
+	expect := map[string]string{"key1": "value1", "key2": "value2", "key3": "value3",
+		"key4": "", "key5": "v5", "key7": "=value7", "k8": "", "k9": "", "space key": "space value"}
+
+	for k, v := range tagKeysToMap(values) {
+		assert.Equal(t, *v, expect[k], "Tag key or value failed parser.")
+	}
+}
+
+func TestDigTags(t *testing.T) {
+	template := "${missing}.${tag}.${pam['item2']['subitem2']['more']}.${pam['item']}.${pam['item2']}." +
+		"${pam['item2']['subitem']}-${pam['item2']['subitem55']}-${pam['item2']['subitem2']['more']}"
```

Comment:
Can you please please make it work without brackets? `{` 

Because if you do, then its exactly the same as the syntax used by rewrite tag which I wanted to replicate: https://docs.fluentbit.io/manual/pipeline/filters/rewrite-tag

Also... I am not sure using brackets is a good idea... fluent bit uses `${}` to reference environment variables in the config file. There's some core code that reads the config file, and then looks for `${Thing}` and then tries to replace it with the value of environment variable `Thing` if it exists.

Reply:
This template library, along with most others require an end tag to find "temptable" items in the string. Very much like the standard `text/template` library uses `{{` and `}}`. Currently the start tag is `${` and the end tag is `}`. These are fully customizable, but we must have some sort of start and end, otherwise the code gets complicated when searching for non-alphanumeric characters to split out the variables. Most variables will be appended or prefixed to dots, slashes and dashes. Using an end-delimiter makes the code a lot cleaner and faster. Faster being the most important aspect. 

I chose the values here because they felt natural; much like `bash`. We can change these values to anything else to avoid other templating engines working on the same data. These come to mind as possible alternatives for start and end tags (but we're not limited to these options):
- `$(` and `)`
- `$<` and `>`
- `$[` and `]`
- `#{` and `}`
- `%{` and `}`
- `#` and `#`

I've also added another commit that modifies how the `tag` value is interpreted. It now allows for up to 10 values instead of the previous iteration only allowing 2 values. By values, I mean the number of dots in the tag. If we find a need for more than 10 values, we can make that work too, but using a single character 0-9 is slightly faster than looking for one or more characters (1-99+), and if we don't surpass 10 dots in a tag name this is sufficient.

Reply:
After testing it became painfully obvious that `${}` will not work, so I changed it to `#()` and this works beautifully. I have this approach working with dynamic log group names and stream names.

Reply:
```
    log_group_name      /ecs/twitch-fluentbit/#(tag)/#(host)
    log_stream_name     #(tag[0]).#(ident)
```

Turns into this:

![Screen Shot 2020-08-19 at 5 34 47 PM](https://user-images.githubusercontent.com/2402929/90703894-dad06c80-e243-11ea-8362-891abb27bc15.png)
![Screen Shot 2020-08-19 at 5 34 35 PM](https://user-images.githubusercontent.com/2402929/90703899-dc9a3000-e243-11ea-8ebb-197a9d5ef639.png)